src/core/inferFullTargets.ts

```diff
@@ -77,112 +81,99 @@ function inferNonListTarget(
-): RangeTarget {
+  target: PartialRangeTargetDesc,
+  previousTargets: PartialTargetDesc[],
+  actionPreferences?: ActionPreferences
+): RangeTargetDescriptor {
   return {
     type: "range",
     excludeAnchor: target.excludeStart ?? false,
     excludeActive: target.excludeEnd ?? false,
     rangeType: target.rangeType ?? "continuous",
     anchor: inferPrimitiveTarget(
-      target.start,
+      target.anchor,
       previousTargets,
       actionPreferences
     ),
     active: inferPrimitiveTarget(
-      target.end,
-      previousTargets.concat(target.start),
+      target.active,
+      previousTargets.concat(target.anchor),
       actionPreferences
     ),
   };
 }
 
 function inferPrimitiveTarget(
-  target: PartialPrimitiveTarget,
-  previousTargets: PartialTarget[],
-  actionPreferences: ActionPreferences
-): PrimitiveTarget {
-  const doAttributeInference = !hasContent(target) && !target.isImplicit;
-
-  const previousTargetsForAttributes = doAttributeInference
-    ? previousTargets
-    : [];
-
-  const maybeSelectionType =
-    target.selectionType ??
-    getPreviousAttribute(previousTargetsForAttributes, "selectionType");
+  target: PartialPrimitiveTargetDesc,
+  previousTargets: PartialTargetDesc[],
+  actionPreferences?: ActionPreferences
+): PrimitiveTargetDescriptor {
+  const hasPosition = !!target.modifiers?.find(
+    (modifier) => modifier.type === "position"
+  );
 
+  // Position without a mark can be something like "take air past end of line"
   const mark = target.mark ??
-    (target.position === "before" || target.position === "after"
-      ? getPreviousMark(previousTargets)
-      : null) ?? {
-      type: maybeSelectionType === "token" ? "cursorToken" : "cursor",
+    (hasPosition ? getPreviousMark(previousTargets) : null) ?? {
+      type: "cursor",
     };
 
-  const position =
-    target.position ??
-    getPreviousPosition(previousTargets) ??
-    actionPreferences.position ??
-    "contents";
-
-  const selectionType =
-    maybeSelectionType ??
-    (doAttributeInference ? actionPreferences.selectionType : null) ??
-    "token";
-
-  const insideOutsideType =
-    target.insideOutsideType ??
-    getPreviousAttribute(previousTargetsForAttributes, "insideOutsideType") ??
-    actionPreferences.insideOutsideType;
-
-  const modifier = target.modifier ??
-    getPreviousAttribute(previousTargetsForAttributes, "modifier") ??
-    (doAttributeInference ? actionPreferences.modifier : null) ?? {
-      type: "identity",
-    };
+  const previousModifiers = getPreviousModifiers(previousTargets);
+  const implicitModifiers = target.isImplicit
+    ? [{ type: "toRawSelection" } as Modifier]
+    : undefined;
+
+  const modifiers =
+    target.modifiers != null
+      ? target.modifiers
+      : implicitModifiers ??
+        previousModifiers ??
+        actionPreferences?.modifiers ??
+        [];
+
+  // TODO Is this really a good solution?
+  // "bring line to after this" needs to infer line on second target
```

Comment:
Seems good enough for now.  I'd leave the comment about "bring line to after this", though.  Otherwise I'd have no clue what was going on here 😅

Reply:
I had no idea either when I started to read the original inference code even though I wrote it initially. Comments on this can of things are quite importante